### PR TITLE
ColumnFilter helper to use into iterator

### PR DESCRIPTION
**Pull request by ajnavarro**
ColumnFilter object that is able to:

- Return true or false to check if we need to filter a specific row from the iteration
- Get a list of matching elements to be able to filter directly using them

### CODE REVIEW

**Comment by ajnavarro** [comment]
@bzz can you check if this filter implementation cover all your needs for BlobsIterator please? Thanks!

File: src/main/scala/tech/sourced/api/util/ColumnFilter.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+trait CompiledFilter {
+  def eval(cols: Map[String, Any]): Option[Boolean]
+
+  def matchingCases: Map[String, Seq[Any]] =
+    this.getMatchingCases.groupBy(_._1).mapValues(_.map(_._2))
+
+  private def getMatchingCases: Array[(String, Any)] =
+    this match {
+      case ef: EqualFilter => Array((ef.name, ef.value))
```
Comment (ajnavarro) [inline src/main/scala/tech/sourced/api/util/ColumnFilter.scala]: we need to have into account the `Not` modifier

Comment (ajnavarro) [reply src/main/scala/tech/sourced/api/util/ColumnFilter.scala]: done.

**Comment by bzz** [comment]
👍 looking into it right now

File: src/main/scala/tech/sourced/api/util/ColumnFilter.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      case IsNotNull(attr) => NotFilter(new EqualFilter(attr, null))
+      case Not(child) => NotFilter(compileFilter(child))
+      case And(l, r) => new AndFilter(compileFilter(l), compileFilter(r))
+      case Or(l, r) => new OrFilter(compileFilter(l), compileFilter(r))
+      case _ => UnhandledFilter()
+    }
+  }
+}
+
+trait CompiledFilter {
```
Comment (erizocosmico) [inline src/main/scala/tech/sourced/api/util/ColumnFilter.scala]: shouldn't we seal it to ensure exhaustiveness when pattern matching over this type?

**Comment by bzz** [comment]
With current API the use case for getting commit hashes will look like:

```scala
  filters.flatMap { filter =>
      filter.matchingCases("hash").flatMap { hash => 
         //
      }
  }.toIterator
```

I'm still a bit confused on i.e how will the eq filter get created say, for the commits.

**Comment by bzz** [comment]
One more thing - it would be really nice to have .toString() defined for CompileFilter.

Similar way to .explain(), but without the truncation - this would be of a great aid while debugging

**Comment by bzz** [comment]
👍 

is there anything left, before merging this guy?

**Pull request merged**